### PR TITLE
feat(my account): addressbook

### PR DESCRIPTION
**Pull request by amolpawar-kibo**

### CODE REVIEW

File: components/dialogs/MyAccount/DeleteAddressConfirmationDialog/DeleteAddressConfirmationDialog.tsx [hunk 1]
```diff
@@ -0,0 +1,66 @@
+import React from 'react'
+
+import { Box, Button, Typography } from '@mui/material'
+import { useTranslation } from 'next-i18next'
+
+import { KiboDialog } from '@/components/common'
+
+interface DeleteAddressConfirmationDialogProps {
+  isOpen: boolean
```
Comment (kibo-chandradeeptalaha) [inline components/dialogs/MyAccount/DeleteAddressConfirmationDialog/DeleteAddressConfirmationDialog.tsx:9]: we don't need this prop

File: lib/helpers/buildAddressProps.ts [hunk 2]
```diff
@@ -0,0 +1,12 @@
+import type { CrAddress, CuAddress } from '@/lib/gql/types'
+
+export const buildAddressProps = (address: CuAddress | CrAddress) => {
```
Comment (kibo-kevinwatts) [inline lib/helpers/buildAddressProps.ts:3]: We haven't been using "helpers" like this. Use getters to get values from data or use a function isolated to the component

File: lib/helpers/buildAddressParams.ts [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+      accountId: accountId,
+    },
+  }
+
+  const updateAddressParams = {
+    ...addressParams,
+    contactId: address?.contact?.id as number,
+  }
+
+  return action === 'update' ? updateAddressParams : addressParams
```
Comment (kibo-kevinwatts) [inline lib/helpers/buildAddressParams.ts:25]: accountId is listed twice in addressParams
Why use 'action' for this helper? if its new, contact id won't exist. can just use that?

File: components/my-account/AddressBook/AddressBook.tsx [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+                  }
+                />
+              )}
+            </Box>
+          </Box>
+          <Divider sx={{ marginTop: '1.75rem', marginBottom: '0.25rem' }} />
+        </Box>
+      ))}
+
+      {isAuthenticated && !isAddNewAddress && !isEditAddress && (
```
Comment (kibo-kevinwatts) [inline components/my-account/AddressBook/AddressBook.tsx:159]: isAuthenticated shouldnt be needed. the my-account page is a protected  route (only auth users can see it)

File: components/my-account/AddressBook/AddressBook.tsx [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+            <Typography variant="h2" fontWeight={'500'}>
+              {t('primary')}
+            </Typography>
+          )}
+          <Box sx={{ display: 'flex', justifyContent: 'space-between' }}>
+            <Box sx={{ display: 'block' }}>
+              <AddressCard {...buildAddressProps(item?.address as CuAddress)} />
+            </Box>
+            <Box sx={{ display: 'flex', flexDirection: 'column' }}>
+              <Link href="" passHref>
```
Comment (kibo-kevinwatts) [inline components/my-account/AddressBook/AddressBook.tsx:132]: If we're not navigating anywhere, why next/link?

**Review by kibo-kevinwatts — CHANGES_REQUESTED** [review]

File: components/dialogs/MyAccount/DeleteAddressConfirmationDialog/DeleteAddressConfirmationDialog.tsx (hunk 1 above)
Comment (amolpawar-kibo) [reply components/dialogs/MyAccount/DeleteAddressConfirmationDialog/DeleteAddressConfirmationDialog.tsx]: Done.

File: components/common/AddressForm/AddressForm.tsx [hunk 6]
```diff
@@ -22,7 +22,7 @@ interface AddressFormProps {
   isUserLoggedIn: boolean
   saveAddressLabel?: string
   setAutoFocus?: boolean
-  checkout: Order | undefined
```
[the remark this answers is not in the document]
Comment (amolpawar-kibo) [reply components/common/AddressForm/AddressForm.tsx]: Resolved in merged PR https://github.com/KiboSoftware/nextjs-storefront/pull/108

File: components/dialogs/MyAccount/DeleteAddressConfirmationDialog/DeleteAddressConfirmationDialog.tsx [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+  isDialogCentered: boolean
+  onClose: () => void
+  onDeleteAddress: () => void
+}
+
+// Component
+const DeleteAddressConfirmationDialog = (props: DeleteAddressConfirmationDialogProps) => {
```
[the remark this answers is not in the document]
Comment (amolpawar-kibo) [reply components/dialogs/MyAccount/DeleteAddressConfirmationDialog/DeleteAddressConfirmationDialog.tsx]: Resolved in merged PR https://github.com/KiboSoftware/nextjs-storefront/pull/108

File: components/my-account/AddressBook/AddressBook.tsx (hunk 4 above)
Comment (amolpawar-kibo) [reply components/my-account/AddressBook/AddressBook.tsx]: Done.

File: lib/helpers/buildAddressParams.ts (hunk 3 above)
Comment (amolpawar-kibo) [reply lib/helpers/buildAddressParams.ts:25]: Done.  accountId is required in both the places as per arguments needed in createCustomerAccountContact mutation

File: components/my-account/AddressBook/AddressBook.tsx (hunk 5 above)
Comment (amolpawar-kibo) [reply components/my-account/AddressBook/AddressBook.tsx]: Done

File: lib/helpers/buildAddressProps.ts (hunk 2 above)
Comment (amolpawar-kibo) [reply lib/helpers/buildAddressProps.ts]: Done. used from addressGetters

**Review by kibo-chandradeeptalaha — CHANGES_REQUESTED** [review]
Please take a look

File: components/my-account/AddressBook/AddressBook.tsx [hunk 8]
```diff
@@ -0,0 +1,222 @@
...
+
+      {!isAddNewAddress && !isEditAddress && (
+        <Button
+          variant="contained"
+          color="inherit"
+          sx={{ maxWidth: '26.313rem' }}
+          onClick={handleNewAddress}
+          fullWidth
+        >
+          <AddCircleOutlineIcon sx={{ paddingRight: '1rem' }} />
```
[the remark this answers is not in the document]
Comment (amolpawar-kibo) [reply components/my-account/AddressBook/AddressBook.tsx]: Done

File: components/my-account/AddressBook/AddressBook.tsx [hunk 9]
```diff
@@ -0,0 +1,229 @@
...
+    }
+  }
+
+  return (
+    <Box data-testid={'address-book-component'}>
+      {addresses?.map((item: CustomerContact, index: number) => (
+        <Box paddingY={1} key={item?.id + 'address'}>
+          {index === 0 && (
+            <Typography variant="h2" fontWeight={'500'}>
+              {t('primary')}
```
Comment (TriptiVermaXP) [inline components/my-account/AddressBook/AddressBook.tsx:149]: Primary - Fontsize and weight are not matching with wireframe.
Edit - Fontsize is not matching with wireframe.
Space between Edit and Delete icon should be little less.

File: components/my-account/AddressBook/AddressBook.tsx [hunk 10]
```diff
@@ -0,0 +1,229 @@
...
+
+  const { t } = useTranslation('common')
+  const { showModal, closeModal } = useModalContext()
+  const [isAddNewAddress, setIsAddNewAddress] = useState<boolean>(false)
+  const [isEditAddress, setIsEditAddress] = useState<boolean>(false)
+
+  const [validateForm, setValidateForm] = useState<boolean>(false)
+  const [isDefaultAddress, setIsDefaultAddress] = useState<boolean>(false)
+
+  const [editAddress, setEditAddress] = useState<CustomerContact>()
```
Comment (TriptiVermaXP) [inline components/my-account/AddressBook/AddressBook.tsx:66]: Please group the states, hooks and methods.

File: components/my-account/AddressBook/AddressBook.tsx [hunk 11]
```diff
@@ -0,0 +1,229 @@
...
+  const handleAddressValidationAndSave = () => setValidateForm(true)
+  const handleSaveAddress = async (address: Address) => {
+    if (address?.contact?.id) {
+      try {
+        const params = buildAddressParams({
+          accountId: user?.id,
+          address,
+          isDefaultAddress,
+          addressType: AddressType.SHIPPING.toLowerCase(),
+        })
```
Comment (TriptiVermaXP) [inline components/my-account/AddressBook/AddressBook.tsx:93]: buildAddressParams code is repetitive in both the blocks if-else.
 Also use  AddressType.SHIPPING constant.
 How we are handing this for billing address?

File: components/my-account/AddressBook/AddressBook.tsx [hunk 12]
```diff
@@ -0,0 +1,229 @@
...
+    setIsAddNewAddress(false)
+    setIsEditAddress(false)
+  }
+
+  const handleConfirmDeleteAddress = (deleteAddressProps: DeleteAddressParams) => {
+    showModal({
+      Component: ConfirmationDialog,
+      props: {
+        onConfirm: () => handleDeleteAddress(deleteAddressProps),
+        contentText: t('are-you-sure-you-want-to-delete-this-address'),
```
Comment (TriptiVermaXP) [inline components/my-account/AddressBook/AddressBook.tsx:128]: This key is too long, can we have some short one eg: delete-address-confirmation or something better.

File: components/my-account/AddressBook/AddressBook.tsx [hunk 13]
```diff
@@ -0,0 +1,229 @@
...
+      closeModal()
+    } catch (error) {
+      console.log('Error: delete saved address from my account', error)
+    }
+  }
+
+  return (
+    <Box data-testid={'address-book-component'}>
+      {addresses?.map((item: CustomerContact, index: number) => (
+        <Box paddingY={1} key={item?.id + 'address'}>
```
Comment (TriptiVermaXP) [inline components/my-account/AddressBook/AddressBook.tsx:146]: Avoid using concatenation, try to use backquote

File: components/my-account/AddressBook/AddressBook.tsx [hunk 14]
```diff
@@ -0,0 +1,229 @@
...
+              {t('primary')}
+            </Typography>
+          )}
+          <Box sx={{ display: 'flex', justifyContent: 'space-between' }}>
+            <AddressCard {...buildAddressProps(item?.address as CuAddress)} />
+            <Stack>
+              <Typography
+                variant="body2"
+                sx={{ cursor: 'pointer' }}
+                onClick={() => handleEditAddress(item as CustomerContact)}
```
Comment (TriptiVermaXP) [inline components/my-account/AddressBook/AddressBook.tsx:158]: Try to use verb/action at the end for all methods.
@kibo-sushant Please suggest if its applicable.

File: __test__/e2e/integration/page-template/MyAccountTemplate/MyAccountTemplate.spec.tsx [hunk 15]
```diff
@@ -35,6 +35,22 @@ jest.mock('@/lib/helpers/tokenizeCreditCardPayment', () => {
...
+    await user.click(addressBook)
+
+    const addNewAddressButton = screen.getByRole('button', { name: 'add-new-address' })
+    await user.click(addNewAddressButton)
+
+    const addressForm = screen.getByTestId('address-form')
+    expect(addressForm).toBeVisible()
+  })
+})
+
```
Comment (TriptiVermaXP) [inline __test__/e2e/integration/page-template/MyAccountTemplate/MyAccountTemplate.spec.tsx:53]: Integration tests are missing for add, update and delete the address.
Also graphql mocks are missing in msw.

File: components/my-account/AddressBook/AddressBook.tsx [hunk 16]
```diff
@@ -0,0 +1,229 @@
...
+import type { Address, ContactForm, DeleteAddressParams } from '@/lib/types'
+
+import type { UpdateCustomerAccountContactDetailsParams } from '@/hooks'
+import type {
+  CrAddress,
+  CuAddress,
+  CustomerAccount,
+  CustomerContact,
+  CustomerContactCollection,
+  Maybe,
```
Comment (TriptiVermaXP) [inline components/my-account/AddressBook/AddressBook.tsx:28]: Remove unused Maybe

**Review by TriptiVermaXP — CHANGES_REQUESTED** [review]
Please take a look review comments.
One thing I notices, ConfirmationDialog is placed inside /Dialogs/MyAccount, shouldn't be directly inside the /Dialogs as it could be used in other places?
@kibo-kevinwatts Please suggest.

File: components/my-account/AddressBook/AddressBook.tsx [hunk 17]
```diff
@@ -0,0 +1,222 @@
...
+  const [validateForm, setValidateForm] = useState<boolean>(false)
+  const [isDefaultAddress, setIsDefaultAddress] = useState<boolean>(false)
+
+  const [editAddress, setEditAddress] = useState<CustomerContact>()
+
+  const { addSavedAddressDetails } = useCreateCustomerAddressMutation()
+  const { updateSavedAddressDetails } = useUpdateCustomerAddressMutation()
+  const { deleteSavedAddressDetails } = useDeleteCustomerAddressMutation()
+
+  const addresses = userAddressGetters.getUserShippingAddress(contacts?.items)
```
[the remark this answers is not in the document]
Comment (amolpawar-kibo) [reply components/my-account/AddressBook/AddressBook.tsx]: Done.

File: components/my-account/AddressBook/AddressBook.tsx (hunk 8 above)
[the remark this answers is not in the document]
Comment (amolpawar-kibo) [reply components/my-account/AddressBook/AddressBook.tsx]: Done.

File: components/my-account/AddressBook/AddressBook.tsx [hunk 18]
```diff
@@ -0,0 +1,222 @@
...
+    <Box data-testid={'address-book-component'}>
+      {addresses?.map((item: Maybe<CustomerContact>, index: number) => (
+        <Box paddingY={1} key={item?.id + 'address'}>
+          {index === 0 && (
+            <Typography variant="h2" fontWeight={'500'}>
+              {t('primary')}
+            </Typography>
+          )}
+          <Box sx={{ display: 'flex', justifyContent: 'space-between' }}>
+            <Box sx={{ display: 'block' }}>
```
[the remark this answers is not in the document]
Comment (amolpawar-kibo) [reply components/my-account/AddressBook/AddressBook.tsx]: Done.

File: components/my-account/AddressBook/AddressBook.tsx [hunk 19]
```diff
@@ -0,0 +1,222 @@
...
+              <AddressCard {...buildAddressProps(item?.address as CuAddress)} />
+            </Box>
+            <Box sx={{ display: 'flex', flexDirection: 'column' }}>
```
[the remark this answers is not in the document]
Comment (amolpawar-kibo) [reply components/my-account/AddressBook/AddressBook.tsx]: Done.

File: components/my-account/AddressBook/AddressBook.tsx [hunk 20]
```diff
@@ -0,0 +1,222 @@
...
+
+  const handleAddressValidationAndSave = () => setValidateForm(true)
+  const handleSaveAddress = async (address: Address) => {
+    if (address?.contact?.id) {
+      try {
+        const params = buildAddressParams({
+          accountId: user?.id,
+          address,
+          isDefaultAddress,
+          addressType: 'shipping',
```
[the remark this answers is not in the document]
Comment (amolpawar-kibo) [reply components/my-account/AddressBook/AddressBook.tsx]: Done.

File: components/my-account/AddressBook/AddressBook.tsx [hunk 21]
```diff
@@ -0,0 +1,222 @@
...
+      await deleteSavedAddressDetails.mutateAsync(deleteAddressProps)
+      closeModal()
+    } catch (error) {
+      console.log('Error: delete saved address from my account', error)
+    }
+  }
+
+  return (
+    <Box data-testid={'address-book-component'}>
+      {addresses?.map((item: Maybe<CustomerContact>, index: number) => (
```
[the remark this answers is not in the document]
Comment (amolpawar-kibo) [reply components/my-account/AddressBook/AddressBook.tsx]: Done.

File: components/my-account/AddressBook/AddressBook.tsx (hunk 9 above)
Comment (amolpawar-kibo) [reply components/my-account/AddressBook/AddressBook.tsx]: Resolved in merged PR https://github.com/KiboSoftware/nextjs-storefront/pull/108

File: components/my-account/AddressBook/AddressBook.tsx (hunk 10 above)
Comment (amolpawar-kibo) [reply components/my-account/AddressBook/AddressBook.tsx:74]: Done.

File: components/my-account/AddressBook/AddressBook.tsx (hunk 11 above)
Comment (amolpawar-kibo) [reply components/my-account/AddressBook/AddressBook.tsx:107]: Done.

File: components/my-account/AddressBook/AddressBook.tsx (hunk 13 above)
Comment (amolpawar-kibo) [reply components/my-account/AddressBook/AddressBook.tsx:160]: Done.

File: components/my-account/AddressBook/AddressBook.tsx (hunk 16 above)
Comment (amolpawar-kibo) [reply components/my-account/AddressBook/AddressBook.tsx]: Done

**Review by TriptiVermaXP — APPROVED** [review]

File: components/my-account/AddressBook/AddressBook.tsx (hunk 12 above)
Comment (amolpawar-kibo) [reply components/my-account/AddressBook/AddressBook.tsx]: Done.

File: __test__/e2e/integration/page-template/MyAccountTemplate/MyAccountTemplate.spec.tsx (hunk 15 above)
Comment (amolpawar-kibo) [reply __test__/e2e/integration/page-template/MyAccountTemplate/MyAccountTemplate.spec.tsx:118]: Done.

File: components/my-account/AddressBook/AddressBook.tsx [hunk 22]
```diff
@@ -0,0 +1,299 @@
...
+const styles = {
+  addNewAddressButtonStyle: {
+    maxWidth: '26.313rem',
+    '& > *:first-child': {
+      fontSize: 'inherit',
+    },
+  },
+}
+const buildAddressProps = (address: CuAddress | CrAddress) => {
+  const { address1, address2, cityOrTown, stateOrProvince, postalOrZipCode } = address
```
Comment (kibo-chandradeeptalaha) [inline components/my-account/AddressBook/AddressBook.tsx:62]: if we are sending only CuAddress as param, we don't need the CrAddress type.

**Review by kibo-chandradeeptalaha — APPROVED** [review]
Looks good. just check this one thing

**Review by kibo-kevinwatts — APPROVED** [review]

**Pull request merged**